File: leetcode_75/hashmap_set/equal_row_and_column_pairs_2352.py
from typing import List
from collections import Counter
class Solution:
    def equalPairs(self, grid: List[List[int]]) -> int:

        counter = 0
        col = [tuple([k[i] for k in grid]) for i in range(len(grid)) ]
        ccol = Counter(col)

        for j in grid:
            counter += ccol[tuple(j)]
        
        return counter

aa = Solution()
aa.equalPairs([[3,1,2,2],
               [1,4,4,4],
               [2,4,2,2],
               [2,5,2,2]])

# key points
# 1. 要變成 tuple 才可以轉變成 hashable，並且tuple不像是set()，不會把重複吃掉
# 2. Counter資料查找，如果沒有，會得到0，不會報錯

File: leetcode_75/hashmap_set/equal_row_and_column_pairs_2352.java
package leetcode_75.hashmap_set;

import java.util.*;

class Solution {
    public int equalPairs(int[][] grid) {
        int n = grid.length;
        int counter = 0;
        Map<List<Integer>, Integer> colMap = new HashMap<>();

        for (int c = 0; c < n; c++) {
            List<Integer> colList = new ArrayList<>();
            for (int r = 0; r < n; r++) {
                colList.add(grid[r][c]);
            }
            colMap.put(colList, colMap.getOrDefault(colList, 0) + 1);
        }

        for (int r = 0; r < n; r++) {
            List<Integer> rowList = new ArrayList<>();
            for (int c = 0; c < n; c++) {
                rowList.add(grid[r][c]);
            }
            counter += colMap.getOrDefault(rowList, 0);
            // 這邊可以查找多次，因為邊是rowList一直刷新，然後丟進具比對，所以可以不會因為重複就沒有加上次數
        }

        return counter;
    }
}
